src/Juvix/Compiler/Tree/Transformation/Validate.hs

```diff
@@ -0,0 +1,267 @@
+module Juvix.Compiler.Tree.Transformation.Validate where
+
+import Juvix.Compiler.Core.Data.BinderList qualified as BL
+import Juvix.Compiler.Tree.Error
+import Juvix.Compiler.Tree.Extra.Base (getNodeLocation)
+import Juvix.Compiler.Tree.Extra.Recursors
+import Juvix.Compiler.Tree.Extra.Type
+import Juvix.Compiler.Tree.Transformation.Base
+
+inferType :: forall r. (Member (Error TreeError) r) => InfoTable -> FunctionInfo -> Node -> Sem r Type
+inferType tab funInfo = goInfer mempty
+  where
+    goInfer :: BinderList Type -> Node -> Sem r Type
+    goInfer bl = \case
+      Binop x -> goBinop bl x
+      Unop x -> goUnop bl x
+      Const x -> goConst bl x
+      MemRef x -> goMemRef bl x
+      AllocConstr x -> goAllocConstr bl x
+      AllocClosure x -> goAllocClosure bl x
+      ExtendClosure x -> goExtendClosure bl x
+      Call x -> goCall bl x
+      CallClosures x -> goCallClosures bl x
+      Branch x -> goBranch bl x
+      Case x -> goCase bl x
+      Save x -> goSave bl x
+
+    goBinop :: BinderList Type -> NodeBinop -> Sem r Type
+    goBinop bl NodeBinop {..} = case _nodeBinopOpcode of
+      IntAdd -> checkBinop mkTypeInteger mkTypeInteger mkTypeInteger
+      IntSub -> checkBinop mkTypeInteger mkTypeInteger mkTypeInteger
+      IntMul -> checkBinop mkTypeInteger mkTypeInteger mkTypeInteger
+      IntDiv -> checkBinop mkTypeInteger mkTypeInteger mkTypeInteger
+      IntMod -> checkBinop mkTypeInteger mkTypeInteger mkTypeInteger
+      IntLt -> checkBinop mkTypeInteger mkTypeInteger mkTypeBool
+      IntLe -> checkBinop mkTypeInteger mkTypeInteger mkTypeBool
+      ValEq -> checkBinop TyDynamic TyDynamic mkTypeBool
+      StrConcat -> checkBinop TyString TyString TyString
+      OpSeq -> do
+        checkType bl _nodeBinopArg1 TyDynamic
+        goInfer bl _nodeBinopArg2
+      where
+        loc = _nodeBinopInfo ^. nodeInfoLocation
+
+        checkBinop :: Type -> Type -> Type -> Sem r Type
+        checkBinop ty1' ty2' rty = do
+          ty1 <- goInfer bl _nodeBinopArg1
+          ty2 <- goInfer bl _nodeBinopArg2
+          _ <- unifyTypes' loc tab ty1 ty1'
+          _ <- unifyTypes' loc tab ty2 ty2'
```

Comment:
I'd suggest to either use `void` or define `unifyTypes'_`

Reply:
Thanks, I forgot how `void` is called in Haskell and somehow the web search didn't turn it up

Reply:
done